File: src/com/javascript-practice/linkList/insertNode.js
class Node {
    constructor(val) {
      this.val = val;
      this.next = null;
    }
  }
  
  const insertNode = (head, value, index) => {

    if(head === null) return head;
    if(head !=null && index === 0) {
        let insert = new Node(value);
        insert.next = head;
        return insert;
    }

    let count = 0;
    let current  = head;
    
    while(current!=null) {
        
        if(count === index - 1) {
            let temp = current.next;
            let insertedName = new Node(value);
            current.next = insertedName;
            current.next.next = temp;
        }
        count ++;
        current = current.next;
    }
    return head;
  };
  const a = new Node("a");
  const b = new Node("b");
  const c = new Node("c");
  const d = new Node("d");
  
  a.next = b;
  b.next = c;
  c.next = d;
  
  // a -> b -> c -> d
  
  let result = insertNode(a, 'x', 2);
  // a -> b -> x -> c -> d  
 while(result!=null) {
     console.log(result.val);
     result  = result.next;
 }

File: src/com/structy/LinkedList/InsertNode.java
package com.structy.LinkedList;

import com.gfg.Searching.BinarySearch;

public class InsertNode {
    public Node insertNode(Node head, String target, int index){
        if(index == 0){
            Node newHead = new Node(target);
            newHead.next = head;
            return newHead;
        }

        Node current  = head;
        int count = 0;
        while (current!=null){
            if(count == index -1){
             Node next  = current.next;

             current.next = new Node(target);
             current.next.next = next;
            }
            count++;
            current = current.next;
        }
        return head;
    }

    public Node insertNodeRec(Node head, String target, int index){

        return helper(head, target, index, 0);
    }
    private Node helper(Node head, String target, int index, int count){
        if(head == null)
            return null;
        if(index == 0){
            Node newHead = new Node(target);
            newHead.next = head;
            return newHead;
        }
        if(count == index -1){
            Node tmp = head.next;
            head.next = new Node(target);
            head.next.next =  tmp;
            return head;
        }
        helper(head.next, target, index, count + 1);
        return head;
    }

    public static void main(String[] args) {
        Node a = new Node("a");
        Node b = new Node("b");
        Node c = new Node("c");
        Node d = new Node("d");

        a.next = b;
        b.next = c;
        c.next = d;

        InsertNode insertNode = new InsertNode();
        Node result = insertNode.insertNodeRec(a, "x", 2);
        Node tmp = result;
        while (tmp!=null){
            System.out.print(tmp.strVal);
            tmp = tmp.next;
        }
    }
}
